main/src/ca/uwaterloo/flix/language/phase/unification/BoolMinimization.scala

```diff
@@ -0,0 +1,607 @@
+      val hoistedTerms = f.foldLeft(baseValue) {
+        case (opt, term) => opt.flatMap(
+          acc => term match {
+            case True => None
+            case False => Some(acc)
+            case Or(terms) => Some(acc ++ terms)
+            case t@Var(_) => Some(acc :+ t)
+            case t@NotVar(_) => Some(acc :+ t)
+            case t@And(_) => Some(acc :+ t)
+          }
+        )
+      }
+      hoistedTerms match {
+        case None => True
+        case Some(Vector()) => False // happens for input like False ∨ False
+        case Some(Vector(term)) => term
+        case Some(terms) => Or(terms)
+      }
+    }
+
+    /**
+      * Transform `f` to an equivalent formula that is a conjunction of
+      * disjunctions of either variables or negated variables.
+      * Ex. `x ∧ (x ∨ y) ∧ (z ∨ ¬y ∨ ¬z)`
+      */
+    def toCNF(f: FormulaNNF): FormulaCNF = {
+      f match {
+        case True => True
+        case False => False
+        case Var(v) => Var(v)
+        case NotVar(v) => NotVar(v)
+        case And(terms) => FormulaCNF.mkAnd(terms.map(toCNF))
+        case Or(terms) =>
+          val (simpleTerms, conjunctions) = split(terms)
+          if (conjunctions.isEmpty) {
+            FormulaCNF.mkOr(simpleTerms)
+          } else {
+            val base: FormulaNNF = if (simpleTerms.isEmpty) False else mkOr(simpleTerms)
+            val newTerms: FormulaNNF = conjunctions.foldLeft(base) {
+              case (acc, and) =>
+                // Transform the formula `acc ∨ ( a_1 ∧ a_2 ∧ ... ∧ a_n)`
+                // to `(a_1 ∨ acc) ∧ (a_2 ∨ acc) ∧ ... ∧ (a_n ∨ acc)`
+                mkAnd(and.terms.map(d => mkOr(d, acc)))
+            }
+            toCNF(newTerms)
+          }
+      }
+    }
+
+    /**
+      * Split terms into simple terms and conjunctions.
+      */
+    def split(terms: Vector[Conjunction]): (Vector[FormulaNNF.SimpleTerm], Vector[FormulaNNF.And]) = {
+      val base: (Vector[FormulaNNF.SimpleTerm], Vector[FormulaNNF.And]) = (Vector.empty, Vector.empty)
+      terms.foldLeft(base) {
+        case ((simples, conjunctions), conjunction) => conjunction match {
+          case True => (simples :+ FormulaNNF.True, conjunctions)
+          case False => (simples :+ FormulaNNF.False, conjunctions)
+          case Var(v) => (simples :+ FormulaNNF.Var(v), conjunctions)
+          case NotVar(v) => (simples :+ FormulaNNF.NotVar(v), conjunctions)
+          case and@And(_) => (simples, conjunctions :+ and)
+        }
+      }
+    }
+
+  }
+
+  private sealed trait FormulaCNF
+
+  private object FormulaCNF {
+
+    sealed trait Disjunction extends FormulaCNF
+
+    /**
+      * A conjunction of disjunctions. And empty list is `True`.
+      */
+    case class And(terms: Vector[Disjunction]) extends FormulaCNF
+
+    /**
+      * A disjunction of simple terms. And empty list is `True`.
+      */
+    case class Or(terms: Set[FormulaNNF.SimpleTerm]) extends Disjunction
```

Comment:
I think this is painting yourself into a corner. Complicated type hierarchies are very difficult to reason about. With algebraic data types you would just define Formula, DnfFormula, and CnfFormula and then define functions directly on them. Yes, you may have to to write the "same function" twice. But that's also how Flix works. Imagine if the ErasedAst was a complicated hierarchy that accounted for all the information from previous compiler phases through a multitude of traits. It would be very difficult to understand and maintain. Instead, Flix is structured as completely separate ADTs and functions mapping between them. We should aim for the same here.